drv/lpc55-update-server/src/images.rs

```diff
@@ -26,79 +45,494 @@ extern "C" {
 // Location of the NXP header
 pub const HEADER_BLOCK: usize = 0;
 
-// NXP LPC55's mixed header/vector table offsets
-const RESET_VECTOR_OFFSET: usize = 0x04;
-pub const LENGTH_OFFSET: usize = 0x20;
-pub const HEADER_OFFSET: u32 = 0x130;
-const MAGIC_OFFSET: usize = HEADER_OFFSET as usize;
+// An image may have an ImageHeader located after the
+// LPC55's mixed header/vector table.
+pub const IMAGE_HEADER_OFFSET: u32 = 0x130;
 
-// Perform some sanity checking on the header block.
-pub fn validate_header_block(
-    component: RotComponent,
-    slot: SlotId,
-    block: &[u8; BLOCK_SIZE_BYTES],
-) -> Result<(), UpdateError> {
-    let exec = image_range(component, slot).1;
+/// Address ranges that may contain an image during storage and active use.
+/// `store` and `exec` ranges are the same except for `stage0next`.
+pub struct FlashRange {
+    pub store: Range<u32>,
+    pub exec: Range<u32>,
+}
+
+/// Get the flash storage address range and flash execution address range.
+pub fn flash_range(component: RotComponent, slot: SlotId) -> FlashRange {
+    // Safety: Linker defined symbols are trusted.
+    unsafe {
+        match (component, slot) {
+            (RotComponent::Hubris, SlotId::A) => FlashRange {
+                store: __IMAGE_A_BASE.as_ptr() as u32
+                    ..__IMAGE_A_END.as_ptr() as u32,
+                exec: __IMAGE_A_BASE.as_ptr() as u32
+                    ..__IMAGE_A_END.as_ptr() as u32,
+            },
+            (RotComponent::Hubris, SlotId::B) => FlashRange {
+                store: __IMAGE_B_BASE.as_ptr() as u32
+                    ..__IMAGE_B_END.as_ptr() as u32,
+                exec: __IMAGE_B_BASE.as_ptr() as u32
+                    ..__IMAGE_B_END.as_ptr() as u32,
+            },
+            (RotComponent::Stage0, SlotId::A) => FlashRange {
+                store: __IMAGE_STAGE0_BASE.as_ptr() as u32
+                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
+                exec: __IMAGE_STAGE0_BASE.as_ptr() as u32
+                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
+            },
+            (RotComponent::Stage0, SlotId::B) => FlashRange {
+                store: __IMAGE_STAGE0NEXT_BASE.as_ptr() as u32
+                    ..__IMAGE_STAGE0NEXT_END.as_ptr() as u32,
+                exec: __IMAGE_STAGE0_BASE.as_ptr() as u32
+                    ..__IMAGE_STAGE0_END.as_ptr() as u32,
+            },
+        }
+    }
+}
+
+/// Does (component, slot) refer to the currently running Hubris image?
+pub fn is_current_hubris_image(component: RotComponent, slot: SlotId) -> bool {
+    // Safety: We are trusting the linker.
+    flash_range(component, slot).store.start
+        == unsafe { &__this_image } as *const _ as u32
```

Comment:
Consider using `addr_of!` for this and the other static address computation expressions --- I think it's clearer, and it's _much_ harder to get wrong.

Reply:
Done